src/main/scala/com/actian/spark_vector/colbuffer/ColumnBuffer.scala

```diff
@@ -22,89 +22,173 @@ import com.actian.spark_vector.colbuffer.singles._
 import com.actian.spark_vector.colbuffer.string._
 import com.actian.spark_vector.colbuffer.time._
 import com.actian.spark_vector.colbuffer.timestamp._
+import com.actian.spark_vector.datastream.padding
+import com.actian.spark_vector.BooleanExpr
 
 import java.nio.ByteBuffer
 import java.nio.ByteOrder
+import java.nio.BufferOverflowException
 
 import scala.reflect.{ ClassTag, classTag }
 import scala.throws._
 
-/** Abstract class to be used when implementing the class for a typed ColumnBuffer
- *  (e.g. object IntColumnBuffer extends ColumnBuffer[Int])
+/**
+ * Abstract class to be used when implementing the class for a typed ColumnBuffer
+ * (e.g. object IntColumnBuffer extends ColumnBuffer[Int])
  *
- *  This class implements the base methods for buffering vectors of column values.
- *  The `put`, `flip`, `clear` methods are according to the Buffer interface.
- *  The column value serialization should be implemented within the concrete
- *  (typed) class instead.
+ * This class implements the base `put` and `get` methods for serializing (IN-type) and
+ * deserializing (OUT-type) vectors of column values.
  *
- *  @param name the column's name
- *  @param maxValueCount the maximum number of values to store within the buffer
- *  @param valueWidth the width of the value's data type
- *  @param alignSize the data type's alignment size
- *  @param nullable whether this column accepts null values or not
+ * @param name the column's name
+ * @param maxValueCount the maximum number of values to store within the buffer
+ * @param valueWidth the width of the value's data type
+ * @param alignSize the data type's alignment size
+ * @param nullable whether this column accepts null values or not
  */
-abstract class ColumnBuffer[@specialized T: ClassTag](name: String, maxValueCount: Int, valueWidth: Int, val alignSize: Int, val nullable: Boolean) {
-  private final val NullMarker = 1: Byte
-  private final val NonNullMarker = 0: Byte
+private[colbuffer] abstract class ColumnBuffer[@specialized IN: ClassTag, @specialized OUT: ClassTag](val name: String,
+  val maxValueCount: Int, val valueWidth: Int, val alignSize: Int, val nullable: Boolean) extends Serializable {
+  val valueTypeIn = classTag[IN]
+  val valueTypeOut = classTag[OUT]
+
+  @inline def put(source: IN, buffer: ByteBuffer): Unit
+
+  @inline def get(buffer: ByteBuffer): OUT
+}
+
+private[colbuffer] sealed abstract class RWColumnBuffer(col: ColumnBuffer[_, _])
+  extends Serializable {
+  protected final val NullMarker = 1: Byte
+  protected final val NonNullMarker = 0: Byte
+
+  val alignSize = col.alignSize
```

Comment:
yes, all of them

I want to limit access to ColumnBuffer objects and its most parameters. Now this class has all params @public because we need them in R/W col buffers, however not all of them are needed outside .colbuffer's package